### PR TITLE
[BE-4328] Fixes convertion of subqueries in on condition of join.

### PR DESCRIPTION
**Pull request by keatoooon**
Associated Bodo PR: https://github.com/Bodo-inc/Bodo/pull/4989

The original PR fixed the issue of incorrect rexInputRefs being created whenever there were sub queries in the joins. However, there were additional issues when converting the value of the subquerry itself that were not caught.

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java
Comment (njriasan) [inline core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java:1349]: To clarify this is because there are potentially multiple keys for the same starting idx?

File: core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java
Comment (njriasan) [inline core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java:3458]: How stable would you say this is? Are you worried this is too hacky?

File: core/src/test/resources/org/apache/calcite/test/SqlToRelConverterTest.xml
Comment (njriasan) [inline core/src/test/resources/org/apache/calcite/test/SqlToRelConverterTest.xml:3479]: To clarify these are only subquery examples that have the wrong output/need changing right.

**Review by njriasan — COMMENTED** [review]
Overall I think this makes sense and I just have a couple questions about how reliable you think this change is. Happy to approve if you think this is stable.

File: core/src/test/resources/org/apache/calcite/test/SqlToRelConverterTest.xml
Comment (keatoooon) [reply core/src/test/resources/org/apache/calcite/test/SqlToRelConverterTest.xml:3479]: Correct

File: core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java
Comment (keatoooon) [reply core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java:3458]: It is stable specifically for scalar sub queries, they ALWAYS result in a RexRangeRef with an offset of 0, it's commented later in the code that handles the substitution. This isn't necessarily stable across all possible sub queries (though none of them worked before either), we'll need an actual plan to get 100% coverage.

Comment (njriasan) [reply core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java:3458]: Okay sounds good. Can you file an issue for working on these changes + a confluence docs adding some additional context?

**Comment by njriasan** [comment]
@keatoooon Do you know why CI is failing? I'd like to wait to approve until after CI is working.

**Comment by keatoooon** [comment]
CI failure was due to a style issue, but the style tester doesn't fail locally, going to re-run and see if it fixes it.

File: core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java
Comment (njriasan) [reply core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java:1349]: I want to make sure this is explained to me before I approve.

File: core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java
Comment (njriasan) [reply core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java:3458]: Can you add the issue and/or confluence doc before we merge this.

**Review by njriasan — COMMENTED** [review]
@keatoooon I have two unaddressed comment but once those are responded to I can approve and we can merge.

File: core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java
Comment (keatoooon) [reply core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java:1349]: So, in the case that we have some expression like "(A, B) in (X, Y)",

The leftkeys will be A and B, and the root of the blackboard will be (outputStartingIdx number of rows), X, Y.

Originally, what was happening here was that we weren't incrementing the inputRef that was being generated for the right hand side of the equality. IE, the condition ended up being:

A == X AND B == X

With this change, we should generate:

A == X and B == Y

File: core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java
Comment (keatoooon) [reply core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java:3458]: linked the issue and added a comment (https://bodo.atlassian.net/browse/BE-4307)

File: core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java
Comment (njriasan) [reply core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java:1349]: Okay this matches my expectations. Thank you!

**Review by njriasan — APPROVED** [review]
LGTM! Thank you @keatoooon

**Pull request merged**
